src/game/client/shadow/w3dbuffermanager.h

```diff
@@ -0,0 +1,166 @@
+/**
+ * @file
+ *
+ * @author tomsons26
+ *
+ * @brief
+ *
+ * @copyright Thyme is free software: you can redistribute it and/or
+ *            modify it under the terms of the GNU General Public License
+ *            as published by the Free Software Foundation, either version
+ *            2 of the License, or (at your option) any later version.
+ *            A full copy of the GNU General Public License can be found in
+ *            LICENSE
+ */
+
+#include "always.h"
+#include <new>
+
+class DX8VertexBufferClass;
+class DX8IndexBufferClass;
+
+class W3DBufferManager
+{
+public:
+    enum VBM_FVF_TYPES
+    {
+        VBM_FVF_XYZ,
+        VBM_FVF_XYZD,
+        VBM_FVF_XYZUV,
+        VBM_FVF_XYZDUV,
+        VBM_FVF_XYZUV2,
+        VBM_FVF_XYZDUV2,
+        VBM_FVF_XYZN,
+        VBM_FVF_XYZND,
+        VBM_FVF_XYZNUV,
+        VBM_FVF_XYZNDUV,
+        VBM_FVF_XYZNUV2,
+        VBM_FVF_XYZNDUV2,
+        VBM_FVF_XYZRHW,
+        VBM_FVF_XYZRHWD,
+        VBM_FVF_XYZRHWUV,
+        VBM_FVF_XYZRHWDUV,
+        VBM_FVF_XYZRHWUV2,
+        VBM_FVF_XYZRHWDUV2,
+        MAX_FVF
+    };
+
+    struct W3DVertexBufferSlot;
+
+    struct W3DRenderTask
+    {
+        W3DRenderTask *m_nextTask;
+    };
+
+    struct W3DVertexBuffer
+    {
+        VBM_FVF_TYPES m_format;
+        W3DVertexBufferSlot *m_usedSlots;
+        int m_startFreeIndex;
+        int m_size;
+        W3DVertexBuffer *m_nextVB;
+        DX8VertexBufferClass *m_DX8VertexBuffer;
+        W3DRenderTask *m_renderTaskList;
+    };
+
+    struct W3DVertexBufferSlot
+    {
+        int m_size;
+        int m_start;
+        W3DVertexBuffer *m_VB;
+        W3DVertexBufferSlot *m_prevSameSize;
+        W3DVertexBufferSlot *m_nextSameSize;
+        W3DVertexBufferSlot *m_prevSameVB;
+        W3DVertexBufferSlot *m_nextSameVB;
+    };
+
+    struct W3DIndexBufferSlot;
+
+    struct W3DIndexBuffer
+    {
+        W3DIndexBufferSlot *m_usedSlots;
+        int m_startFreeIndex;
+        int m_size;
+        W3DIndexBuffer *m_nextIB;
+        DX8IndexBufferClass *m_DX8IndexBuffer;
+    };
+
+    struct W3DIndexBufferSlot
+    {
+        int m_size;
+        int m_start;
+        W3DIndexBuffer *m_IB;
+        W3DIndexBufferSlot *m_prevSameSize;
+        W3DIndexBufferSlot *m_nextSameSize;
+        W3DIndexBufferSlot *m_prevSameIB;
+        W3DIndexBufferSlot *m_nextSameIB;
+    };
+
+    enum
+    {
+        MAX_NUMBER_SLOTS = 4096,
```

Comment:
This looks like the number that gets exceeded in game.

Reply:
It is a but it doesn't mean anything in the context of this function, volumetric shadow classes are needed to understand how it occurs